ch4/ofi: enabling multiple vci for pt2pt operations

hzhou opened the pull request:
## Pull Request Description

We'll enable multiple vci pt2pt operations in the ofi netmod path. 

* We'll use a default communicator based vci method for the basic use case that an application duplicates multiple comm, one for each thread. In this case, each comm will get to assigned a different vci in a round robin fashion.

* Progress will poll each vci in turn. This is a temporary patch. It should pose no impact for the backward case when multiple vci is not enabled. For multiple-vci case when both it is configured and run-time cvar enabled, it will pose as global contention point and provide slower performance (even than single-vci). A coming PR will address this performance issue.

* Move the lock in send/recv paths from ch4-layer down to netmod/shmmod layer. This allows shmmod, netmod, as well as active message path all may use different vci schemes. For now, we are only enabling the pt2pt patch in OFI netmod.

* Since we moved the critical section down, some critical code at ch4-layer is exposed. In particular, the any-source receive and internal cancel receive path need coordination from both shmmod and netmod. Because both modules may take the same or different vci locks, we allow recursive locking in a few places for simplicity.

## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/warnings/custom --with-ch4-max-vcis=8

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most

hzhou commented:
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
@raffenet The tests are mostly clear now. I'll investigate the two sporadic time outs -- `./threads/comm/dup_leak_test 2` and `./threads/spawn/th_taskmaster 2`, which I believe are not introduced in this PR. Meanwhile, this PR is ready for your review.

hzhou commented:
@raffenet Now that this PR has grown into `37` commits, please let me know if you'd like me to split into a few per-requisite PRs.

raffenet commented on src/mpi/coll/igather/igather_inter_sched_short.c:
```diff
@@ -26,7 +26,10 @@ int MPIR_Igather_inter_sched_short(const void *sendbuf, int sendcount, MPI_Datat
     local_size = comm_ptr->local_size;
 
-    if (root == MPI_ROOT) {
+    if (root == MPI_PROC_NULL) {
```
All processes need to specify the same root value. Supplying `MPI_PROC_NULL` surely is not allowed by the MPI standard, is it?

raffenet replied:
Oh this is inter-communicator, now I see.

hzhou replied:
Yeah, just for the record/information, here is the quote from standard:
```
For this, the root process uses the special root value MPI_ROOT; all other processes in the same group as the root use MPI_PROC_NULL.
```

raffenet commented on src/mpid/ch4/src/ch4_probe.h:
```diff
@@ -127,13 +126,12 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_improbe_safe(int source,
     MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI(0).lock);
-
+#endif
+    mpi_errno = MPIDI_improbe_unsafe(source, tag, comm, context_offset, av, flag, message, status);
```
I've gotten as far as these patches to move locks to the lower layers. It seems now the "safe" functions just call "unsafe" APIs without any protection. I.e. the unsafe functions are actually safe? Either I'm missing something, or the naming is going to need to be cleaned up.

hzhou replied:
Yes, they can be cleaned up. I am thinking doing that separately as that potentially creates more code conflicts for the down streams.

raffenet commented:
> @raffenet Now that this PR has grown into `37` commits, please let me know if you'd like me to split into a few prerequisite PRs.

I think some split up would be good. Maybe try to limit to 10-15 commits max per PR.

hzhou commented:
> > @raffenet Now that this PR has grown into `37` commits, please let me know if you'd like me to split into a few prerequisite PRs.
> 
> I think some split up would be good. Maybe try to limit to 10-15 commits max per PR.

Sounds good. I'll do that.

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8 --disable-ofi-domain
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

wesbland commented on src/mpid/ch4/src/ch4_recv.h:
```diff
@@ -317,12 +317,10 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_cancel_recv_safe(MPIR_Request * rreq)
 
-    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
-
+    int vci = MPIDI_Request_get_vci(rreq);
```
Typo in the commit message here:

```diff
-ch4: take the correct vci lock in cancle recv
+ch4: take the correct vci lock in cancel recv
```

hzhou replied:
Thanks.

wesbland commented on src/mpid/ch4/netmod/ofi/ofi_send.h:
```diff
@@ -345,14 +350,29 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
+            vni_dst = MPIDI_vci_get_dst(comm, rank, tag) % MPIDI_OFI_global.num_vnis; \
+        } \
+    } while (0)
+
```
Ugh... we spent a lot of time unrolling Charles's confusing no-argument macros. Please don't add more. Can this either be a `static inline` or at the very least, a macro with arguments to make it more clear what this is doing? When I was trying to read through the code, it really looked like `vni_src` was undefined.

hzhou replied:
I understand. I thought I was defining and using this macro locally, within context it can be tolerated. Will adding `vni_src`, `vni_dst` to the macro argument make it better?

I haven't quite decided how optimized and how the workq code will finalize and having a macro avoids the potential typos and inconsistency which is often difficult to hunt down.

wesbland replied:
Adding `vni_src` and `vni_dst` would clean things up enough for me.

hzhou replied:
done.

wesbland commented on src/mpid/ch4/netmod/ofi/ofi_probe.h:
```diff
@@ -98,6 +98,14 @@ static inline int MPIDI_OFI_do_iprobe(int source,
+        vni_src = MPIDI_vci_get_src(comm, comm->rank, tag) % MPIDI_OFI_global.num_vnis; \
+        vni_dst = MPIDI_vci_get_dst(comm, comm->rank, tag) % MPIDI_OFI_global.num_vnis; \
+    } while (0)
+
```
Same thing about macros here.

hzhou replied:
done (also in ofi_recv).

wesbland commented:
Obviously, that's not a full (or even good 😄 ) review of the code, but there were a few things I noticed in a quick glance. @sagarth / @shawnccx have more useful opinions here than me.

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/warnings/custom --with-ch4-max-vcis=8
netmod: ch4:ofi

hzhou commented:
test:mpich/warnings/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
> I said before the safe/unsafe naming will need to get cleaned up at some point, which still holds.

I don't disagree. Do you prefer to do it in this PR or a separate PR?

raffenet commented:
> > I said before the safe/unsafe naming will need to get cleaned up at some point, which still holds.
> 
> I don't disagree. Do you prefer to do it in this PR or a separate PR?

Honestly, I don't have a strong preference. It was really more of a note for the additional reviewers.

sagarth commented:
@raffenet We'll review this soon.

hzhou commented on src/mpid/ch4/netmod/ofi/ofi_probe.h:
```diff
@@ -111,14 +111,19 @@ static inline int MPIDI_NM_mpi_improbe(int source,
 
 #ifdef MPIDI_ENABLE_LEGACY_OFI
     if (!MPIDI_OFI_ENABLE_TAGGED) {
+        MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
```
The `workq` config is orthogonal to `am-only` (or `legacy-ofi`) config, at least in master. Does `workq` works for `am-only` config in Intel's branch?

hzhou replied:
Actually I with draw my previous comments. I guess the key here is `workq` lives in `ch4-layer`. At netmod layer, there are two branches: if there is a request passed down, we need to take the vci from the request. Otherwise, we create request from vci determined by netmod.

Currently, mulitvci in this PR only works with the latter, i.e. without workq. So I guess we need add an assertion that workq and multi-vci cannot be active simultaneously.

Separate work is needed to make multi-vci work with workq -- as well as am, rma -- so not in this PR.

sagarth replied:
The request based solution that you described would help if we need to take a lock. However, for workq, I think we may not even need to take any lock at the netmod level.  I'm not sure if the current design supports this part.

hzhou replied:
Current master does not support lockless mode. The netmod need ensure consistent locking -- for example, if send sends to a wrong vni, the recv end may never recv it. Therefore, it is difficult to let workq decide the vnis at ch4 layer. We may expose the netmod vni mapping function -- which adds the netmod api, probably shmmod api as well. I would like to explore the solution of using child request similar to persistent request for workq first. Anyway, workq will need separately work.

hzhou replied:
> Where would we make that decision after this PR gets merged into the master? 

It is the same decision, but moved one layer down.

Actually I am not sure how current Intel branch would work. If you decide at ch4-layer to take lockless, that means you can't have active message and shmem since both will need lock?

sagarth replied:
Currently, lockless is applicable only to ofi-netmod. Sure, deciding at ofi-netmod will suit better under the new design change. Do we expose the multi-threading models to ofi-netmod, so that it can make decisions based on what is used?

hzhou replied:
Everything in the ch4-layer is exposed to the netmod, right?  Separately, since the lockless is ofi-only, we probably should change that config option as an ofi netmod option, for example, `--enable-ch4-ofi-lockless`?

sagarth replied:
We should be able to use the things from ch4 at netmod. Yea, moving the lockless setup to the ofi level may be fine or fit well because the locking is moving there too.

hzhou commented on src/mpid/ch4/netmod/ofi/ofi_events.c:
```diff
@@ -186,10 +186,15 @@ static int recv_event(struct fi_cq_tagged_entry *wc, MPIR_Request * rreq, int ev
-        MPIDI_OFI_CALL_RETRY(fi_tinjectdata(MPIDI_OFI_global.ctx[0].tx, NULL /* buf */ ,
+        /* NOTE: use target rank, reply to src */
+        int vni_src =
+            MPIDI_vci_get_src(c, c->rank, rreq->status.MPI_TAG) % MPIDI_OFI_global.num_vnis;
```
Are you suggesting adding OFI wrappers -- `MPIDI_OFI_vci_get_src` and `MPIDI_OFI_vci_get_dst`?

sagarth replied:
Yea such wrapper should work good.

hzhou replied:
done

hzhou commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -123,7 +125,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
 #endif
         {
-            MPIDI_OFI_REQUEST_CREATE(*request, MPIR_REQUEST_KIND__RECV, 0);
+            MPIDI_OFI_REQUEST_CREATE(*request, MPIR_REQUEST_KIND__RECV, vni_dst);
```
`MPID_Recv_init` creates persistent request, which will never enter netmod, so having it in `vci=0` is fine.

sagarth replied:
Oh that's right. We create new request objects at issue time.

hzhou commented:
Rebased. Added commits to add `MPIDI_OFI_get_vni_src/dst` wrappers. Added assertion that workq only works with single vni for now.

@raffenet I would like to leave the "lock move" and "enable" commits separate. This is because not all paths are moved or enabled yet. Having them separately makes it easier to check and for future work on other paths.

test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4
testlist: threads

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/ch4/most
EXTRA_CONFIG: --with-ch4-max-vcis=8 --enable-thread-cs=per-vci

hzhou commented:
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
test:mpich/ch4/most

hzhou commented:
Rerun to confirm:
test:mpich/custom --with-ch4-max-vcis=4 --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

hzhou commented:
@raffenet All tests cleared after merging the latest AM patches in master.
